chicken.1
.\" dummy line
.TH CHICKEN 1 "10 Sep 2002"

.SH NAME

chicken \- A Scheme\-to\-C translator

.SH SYNOPSIS

.B chicken
.I PATHNAME
[
.I OPTION ...
]

.SH DESCRIPTION

.I CHICKEN
is a compiler and interpreter for the programming language
.I Scheme
supporting most of the features as described in the
.I Revised^5 Report on
.I the Algorithmic Language Scheme
\.
The
.B chicken
program is the basic Scheme-to-C translator used in this system.
For a more convenient interface, see the manual page for csc(1).

.SH OPTIONS

Enter
.B chicken\ \-help
for a list of supported command-line options.

.SH ENVIRONMENT\ VARIABLES

.TP
.B CHICKEN_PREFIX
Is used as a prefix directory for support files, include-files and libraries.

.TP
.B CHICKEN_INCLUDE_PATH
Contains one or more pathnames where the compiler should additionally look for
include-files, separated by
.B \;
characters.

.TP
.B CHICKEN_OPTIONS
Holds a string of default compiler options that should apply to every invocation of
.B chicken
\.

.SH DOCUMENTATION

More information can be found in the
.I CHICKEN\ User's\ Manual

.SH AUTHORS

The CHICKEN Team.

.SH SEE ALSO

.BR csc(1)
.BR chicken-bug(1)


NEWS
HEAD

- Runtime system
  - deprecated "[+-]nan", "[+-]inf" and other notations "accidentally"
    accepted by Chicken due to the underlying C library's strtod() function,
    standardizing on "[+-]nan.0" and "[+-]inf.0" from R6RS (and soon R7RS),
    when displaying numbers only these forms are generated now.

4.7.3

- Build system
  - version information has been moved into a separate unit to make the
    build-version shown in the banner and accessible through library
    procedures more accurate, this also speeds up the build on version-
    changes (thanks to Jim Ursetto for contributing this)

- Runtime system
  - fixed handling of "inf" and nan" floating-point predicates for Solaris
    (thanks to Claude Marinier)
  - support for re-loading of compiled files has now been completely 
    removed

- Core libraries
  - added "foldl" and "foldr" procedures, which are more efficient and
    more have a more consistent argument order than the corresponding
    SRFI-1 procedures
  - "shuffle" has been deprecated
  - added "queue-length"
  - "queue->list" allocates and returns a fresh list now
  - invoking a parameter-procedure with an argument will return the new
    value
  - added new procedure "quit"
  - port-procedures now check correctly for argument-ports being open 
    (thanks to Peter Bex)
  - irregex bugfixes (by Peter Bex)
  - "repl" accepts an optional evaluator procedure (suggested by John 
    Cowan)

- Core syntax
  - "parameterize" now correctly omits invoking the guard procedure when
    the old value is restored (thanks to Joo ChurlSoo)
  - "optional", "let-optionals" and "let-optionals*" do not anymore check
    for surplus arguments
  - added ":", "the" and "assume" syntax for declaring types
  - added "define-specialization" form to declare type-driven procedure
    call rewrites in compiled code

- Compiler
  - added "-specialize" option and "specialize" declaration which enables
    optimizations of uses of core library procedures based on type-
    information gathered during flow analysis
  - "-optimize-level 3" and higher now implies "-specialize"
  - added option "-strict-types" and "-emit-type-file"
  - progress-information is now only shown with "-debug p"; the "-verbose"
    option only shows informational but noncritical messages
  - fixed several bugs in the unboxing pass
  - added optimizations for some library procedures
  - variable propagation for global variable accesses is done in certain
    situations now
  - the algorithmic complexity of the closure-conversion pass has been
    reduced which speeds up compilation noticably for large files
  - the "-uses" option handles whitespace in unit lists given on the
    command line (thanks to Santosh Rajan)
  - the alternative branch in a conditional is dropped when the condition
    is known to be a non-boolean value (suggested by Joerg Wittenberger)
  - implemented numerous fixes and improvements for flow analysis

- Interpreter
  - ",q" now leaves the currently active REPL invocation instead of
    terminating the process (as suggested by John Cowan)

- Syntax expander
  - fixed devious bug in the invocation and generation of transformer
    procedures
  - using normal "lambda" forms as transformers for local or global
    syntax definitions is deprecated - one should use "syntax-rules",
    "er-macro-transformer" or "ir-macro-transformer" from now on

- Core tools
  - "chicken-profile"
    - fixed some bugs in the profiler and the runtime support code for
      profiling (thanks to Sven Hartrumpf)
    - fixed broken percentage calculation (thanks to "megane")
  - "chicken-status"
    - the "pattern" argument is now actually treated as a pattern and not
      as a regex
  - "chicken-install"
    - added support for "or"-dependencies where a dependency is considered
      installed if one of a set of candidates is available


4.7.2

- Core tools
  - "chicken-install"
    - Added option "-override", which allows storing extension-
      versions in a file - this may be useful to ensure specific 
      versions of a complete set of eggs to be installed
    - Added option "-keep-installed"/"-x" that only installs explicitly
      named extensions given on the command line, if they are not
      already installed
    - Added option "-list" that lists extensions available remotely
    - Added option "-csi" to specify what interpreter should be used
      to execute the setup-script of an installed extension
    - Added option "-scan" that scans a local directory for the highest
      available extension versions available
    - Added option "-reinstall" that reinstalls all currently installed
      eggs, keeping their versions, if possible
    - Fixed bug with "chicken-install -test" when multiple extensions where
      given on the command line (thanks to Kon Lovett)
  - "chicken-status"
    - Added option "-list" that dumps versions of installed extensions
      in a format suitable for "chicken-install -override ..."
  - "csc"
    - The environment-variables "CHICKEN_C_INCLUDE_PATH" and
      "CHICKEN_C_LIBRARY_PATH" can now be used to add default include-
      and link-directories to C-compiler invocations
  - "csi"
    - ",m" now properly switches between modules (thanks to Christian Kellermann)

- Module system
  - Added "interfaces", named groups of exports
  - Added "functors", which are parameterized modules
  - Modules can now be aliased
  - New syntax:
    (define-interface NAME EXPORTS)
    (functor (NAME ...) EXPORTS ...)
  - Extended syntax of "module" for aliasing and functor-instantiation

- Core libraries
  - Cleaned up set of available internal SRFI-feature-identifiers
  - Fixed bugs in "make-pathname" and "normalize-pathname" (thanks to Jim Ursetto)
  - The reader is now more consistent when parsing symbol- and keyword names 
    that are fully or partially escaped (thanks to Kon Lovett)
  - The printer now does a better job at escaping certain single-character
    symbol names

chicken.scm
;;;; chicken.scm - The CHICKEN Scheme compiler (loader/main-module)
;
; Copyright (c) 2008-2011, The Chicken Team
; Copyright (c) 2000-2007, Felix L. Winkelmann
; All rights reserved.
;
; Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following
; conditions are met:
;
;   Redistributions of source code must retain the above copyright notice, this list of conditions and the following
;     disclaimer. 
;   Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following
;     disclaimer in the documentation and/or other materials provided with the distribution. 
;   Neither the name of the author nor the names of its contributors may be used to endorse or promote
;     products derived from this software without specific prior written permission. 
;
; THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
; AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR
; CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
; CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
; SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
; THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
; OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
; POSSIBILITY OF SUCH DAMAGE.


(declare
  (uses chicken-syntax chicken-ffi-syntax 
	srfi-1 srfi-4 utils files extras data-structures support
	compiler optimizer unboxing compiler-syntax scrutinizer driver platform backend 
	srfi-69))


(include "compiler-namespace")
(include "tweaks")


;;; Prefix argument list with default options:

(define compiler-arguments
  (append
   (remove 
    (lambda (x) (string=? x ""))
    (string-split (or (get-environment-variable "CHICKEN_OPTIONS") "")))
   (cdr (argv))))


;;; Process command-line options:
; 
; - remove runtime-options ("-:...")
; - filter out source-filename
; - convert options into symbols (without the initial hyphens)

(define (process-command-line args)
  (let loop ((args args) (options '()) (filename #f))
    (if (null? args)
	(values filename (reverse options))
	(let* ((arg (car args))
	       (len (string-length arg))
	       (char0 (string-ref arg 0)) )
	  (if (and (char=? #\- char0) (> len 1))
	      (if (and (> len 1) (char=? #\: (string-ref arg 1)))
		  (loop (cdr args) options filename)
		  (loop (cdr args) (cons (string->symbol (substring arg 1 len)) options) filename) )
	      (if filename
		  (loop (cdr args) (cons arg options) filename)
		  (loop (cdr args) options arg) ) ) ) ) ) )


;;; Run compiler with command-line options:

(receive (filename options) ((or (user-options-pass) process-command-line) compiler-arguments)
  (let loop ((os options))
    (unless (null? os)
      (let ((o (car os))
	    (rest (cdr os)) )
	(cond ((eq? 'optimize-level o)
	       (let ((level (string->number (car rest))))
		 (case level
		   ((0) 
		    (set! options
		      (cons* 'no-compiler-syntax 'no-usual-integrations options)) )
		   ((1)
		    (set! options (cons 'optimize-leaf-routines options)) )
		   ((2)
		    (set! options 
		      (cons* 'optimize-leaf-routines
			     'inline
			     'unboxing
			     options)) ) 
		   ((3)
		    (set! options
		      (cons* 'optimize-leaf-routines
			     'inline
			     'inline-global
			     'unboxing 'local
			     'specialize
			     options) ) )
		   ((4)
		    (set! options
		      (cons* 'optimize-leaf-routines
			     'inline 
			     'inline-global
			     'unboxing 
			     'specialize
			     'local 'unsafe
			     options) ) )
		   (else
		    (when (>= level 5)
		      (set! options 
			(cons* 'disable-interrupts 
			       'no-trace
			       'unsafe
			       'block
			       'optimize-leaf-routines
			       'no-lambda-info
			       'specialize
			       'optimize-leaf-routines 
			       'no-lambda-info
			       'inline
			       'inline-global
			       'unboxing
			       options) ) ) ) )
		 (loop (cdr rest)) ) )
	      ((eq? 'debug-level o)
	       (let ((level (string->number (car rest))))
		 (case level
		   ((0) (set! options (cons* 'no-lambda-info 'no-trace options)))
		   ((1) (set! options (cons 'no-trace options)))
		   (else (set! options (cons 'scrutinize options))))
		 (loop (cdr rest)) ) )
	      ((memq o valid-compiler-options) (loop rest))
	      ((memq o valid-compiler-options-with-argument)
	       (if (pair? rest)
		   (loop (cdr rest))
		   (quit "missing argument to `-~s' option" o) ) )
	      (else
	       (warning 
		"invalid compiler option (ignored)" 
		(if (string? o) o (conc "-" o)) )
	       (loop rest) ) ) ) ) )
  (apply compile-source-file filename options)
  (exit) )
